optional-goi-analysis/goi-calculations.R

```diff
@@ -0,0 +1,301 @@
+  library(ComplexHeatmap)
+  library(SingleCellExperiment)
+})
+
+#### Read in data --------------------------------------------------------------
+
+goi_list <- data.table::fread(opt$input_goi_list, stringsAsFactors = FALSE)
+sce <- read_rds(file.path(opt$sce))
+
+#### Perform mapping -----------------------------------------------------------
+
+if (opt$perform_mapping == TRUE) {
+  # turn column name into symbol for pulling the column info out of data frame
+  identifier_column_name <- rlang::sym(opt$identifier_column_name)
+  ids_for_mapping <- goi_list %>%
+    dplyr::pull(identifier_column_name)
+  
+  # Define the annotation packages based on the specified organism
+  annotation_list <- list(
+    'Homo sapiens' = "org.Hs.eg.db",
+    'Mus musculus' = "org.Mm.eg.db",
+    'Danio rerio' = "org.Dr.eg.db"  #,
+    # new organisms would go here
+  )
+  # Error handling
+  if (!(opt$organism %in% names(annotation_list))) {
+    stop(paste(opt$organism, "is not supported"))
+  }
+  # Load library and assign annotation package
+  library(annotation_list[[opt$organism]], character.only = TRUE)
+  annotation_package <-
+    eval(parse(text = annotation_list[[opt$organism]]))
+  
+  # Perform mapping
+  goi_list <- mapIds(
+    # organism annotation package
+    annotation_package,
+    # the provided gene identifiers
+    keys = ids_for_mapping,
+    # the type of provided gene identifiers
+    keytype = opt$input_identifiers,
+    # the type of gene identifiers to map to
+    column = opt$output_identifiers,
+    multiVals = opt$multi_mappings
+  ) %>%
+    tibble::enframe(
+      name = opt$identifier_column_name,
+      value = tolower(opt$output_identifiers)
+    ) %>%
+    # enframe() makes a `list` column; we will simplify it with unnest()
+    # This will result in one row of our data frame per list item
+    tidyr::unnest(cols = tolower(opt$output_identifiers)) %>%
+    # grab only the unique rows
+    dplyr::distinct() %>%
+    # join the remaining columns
+    dplyr::left_join(goi_list, by = tolower(opt$identifier_column_name))
+  
+  # Save mapped object to file
+  write_tsv(goi_list, file.path(
+    opt$output_directory,
+    paste0(opt$library_id, "_mapped_genes.tsv")
+  ))
+}
+
+#### Prepare data for plotting -------------------------------------------------
+
+identifier_column_name_sym <- rlang::sym(opt$identifier_column_name)
+
+# get logcounts from normalized sce object
+normalized_sce_logcounts_matrix <- as.matrix(t(logcounts(sce)))
+
+# we need to filter using the column with Ensembl ids
+if(!is.null(opt$output_identifiers) && (opt$output_identifiers == "ENSEMBL")) {
+  ensembl_id_column <- tolower(opt$output_identifiers)
+} else {
+  ensembl_id_column <- opt$identifier_column_name
+}
+
+# filter counts matrix to only data associated with the provided genes of
+# interest
+normalized_sce_logcounts_matrix <- normalized_sce_logcounts_matrix[,colnames(normalized_sce_logcounts_matrix) %in% goi_list[[ensembl_id_column]]]
```

Comment:
```suggestion
normalized_sce_logcounts_matrix <- normalized_sce_logcounts_matrix[,colnames(normalized_sce_logcounts_matrix) %in% goi_rownames]
```
simplifying a little with my previous suggestion